### PR TITLE
Bacnet pics

### PR DESCRIPTION
**Pull request by JoelRedstone**
Original PR from grafnu: https://github.com/grafnu/daq/pull/154
References issue: https://github.com/faucetsdn/daq/issues/130

### CODE REVIEW

**Review by InBrewJ — CHANGES_REQUESTED** [review]
@JoelRedstone FYI - test aux.out is currently corrupted - we need to fix this first before merging this.

https://github.com/arupiot/daq/blob/master/testing/test_aux.out

(This PR should fix it: https://github.com/arupiot/daq/pull/10)

**Comment by JoelRedstone** [comment]
@InBrewJ PTAL

File: testing/test_aux.out [hunk 1]
```diff
@@ -7,8 +7,11 @@ Lint checks
 cmd/inbuild exit code 0
 Extended tests
 RESULT skip protocol.bacnet.version
-RESULT skip protocol.bacnet.version
-RESULT info protocol.bacnet.version Protocol version: 1
+RESULT skip protocol.bacnet.pic
+RESULT info protocol.bacnet.version
```
Comment (InBrewJ) [inline testing/test_aux.out]: @JoelRedstone Should this test not include the change that prints out the protocol version?

**Review by InBrewJ — CHANGES_REQUESTED** [review]

File: testing/test_aux.out (hunk 1 above)
Comment (JoelRedstone) [reply testing/test_aux.out]: When Francesco added the description after the result he forgot to remove a '**\n**' before the description itself. The description it's been printed but just a line below and it's not been picked up by Daq. Fixing it right now. I'm wondering how it's working fine on bacnet.version

Comment (InBrewJ) [reply testing/test_aux.out]: Do you mean that `Protocol version: 1` **shouldn't** be on the same line as `RESULT info protocol.bacnet.version`?

If so, I'm pretty sure they **should** be on the same line (in this context, line 11). @pisuke is this so?

Comment (JoelRedstone) [reply testing/test_aux.out]: **Yes they need to be on the same line**.. That's what I was trying to say

Comment (InBrewJ) [reply testing/test_aux.out]: ahhh, I get you. Good spot! Apologies for the misunderstanding

File: docs/device_report.md [hunk 2]
```diff
@@ -128,6 +129,9 @@ RESULT fail connection.mac_oui
 
 ```
 RESULT skip protocol.bacnet.version
+Bacnet device not found.
```
Comment (InBrewJ) [inline docs/device_report.md]: https://github.com/arupiot/daq/blob/master/docs/device_report.md

The `cloud.udmi.pointset` here (from Trevor) - has the test info on the same line as the result. Is there a reason why 'Bacnet device not found.' etc is on a new line? @pisuke @JoelRedstone tell me if I've fallen behind with the new line/not new line saga

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Analog.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                    add(bacnetObjectType, PropertyIdentifier.priorityArray, encodable);
+                }
+                break;
+
+                default:
+                System.out.println(objectProperty + " not found.");
+        }
+    }
+
+    private void add(BACnetObject bacnetObjectType, PropertyIdentifier propertyIdentifier, Encodable encodable) {
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Analog.java]: add -> addProperty?

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+                minimumOnTime = Integer.parseInt(propertyValue);
+                encodable = new UnsignedInteger(minimumOnTime);
+                add(bacnetObjectType, PropertyIdentifier.minimumOnTime, encodable);
+                break;
+            case "Minimum_Off_Time":
+                minimumOffTime = Integer.parseInt(propertyValue);
+                encodable = new UnsignedInteger(minimumOffTime);
+                add(bacnetObjectType, PropertyIdentifier.minimumOffTime, encodable);
+                break;
+            case "Feeback_Value":
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java:219]: Does this switch need a `default` - just in case?

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                feedbackValue = Integer.parseInt(propertyValue);
+                encodable = new BinaryPV(feedbackValue);
+                add(bacnetObjectType, PropertyIdentifier.feedbackValue, encodable);
+                break;
+        }
+    }
+
+    private void add(BACnetObject bacnetObjectType, PropertyIdentifier propertyIdentifier, Encodable encodable) {
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java]: Is this the same function as in Analog.java?

File: subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            absolute_path += path_arr[count] + "/";
+        }
+        return absolute_path;
+    }
+
+    public String getCSVPath() {
+        return "/config/type";
+    }
+
+      // For debugging
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java]: You could have a debug flag and turn it on and off rather than relying on commented out/uncommented out code

File: subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  String write = "W";
+  String optional = "O";
+  String Supported = "TRUE";
+  String formatProperty = "%-35s%-35s%-15s%-25s";
+
+  Multimap<String, String> result = ArrayListMultimap.create();
+
+  boolean testPassed = true;
+
+  public void validate(String bacnetObjectType, String bacnetObjectProperty, String conformanceCode,
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java:21]: I haven't functionally tested this but:

`testPassed = true` as default.

If anything happens that means your `if`s aren't hit then we may get a false positive. It could be an idea to have the testPassed = false as default - this might be better for debugging especially with a reason for the failure in the device report

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/EntryPoint.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+            throw e;
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+    }
+
+    private static JSONArray readJSONFile() {
+        String jsonFile = "";
+        if (testCase.equals("pass")) {
+            jsonFile = "Faux-Device-Pass.json";
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/EntryPoint.java]: Can we pass these json filenames as arguments somewhere? You might find that the filename changes in the future and you're having to unnecessarily rebuild docker containers...

**Review by InBrewJ — CHANGES_REQUESTED** [review]

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java (hunk 4 above)
Comment (JoelRedstone) [reply subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java:219]: What switch? Sorry i can only see "minimumOffTime" from the code snippet.

File: subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java (hunk 7 above)
Comment (JoelRedstone) [reply subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java:21]: So the idea is to set testPassed to "true" initially. Once one of the if statement get hit (and has a condition that make the test fail) the "testPassed" boolean is changed to "false" making the overall test to fail (since there is no testPassed = true anywhere).  

I have tested this code with different test cases and has always worked fine. But I get your point and you're definitely right (this is code, anything can happen).

Maybe the thing that is missing is an "else" statement at the very end to make the test fail if none of the if statement are hit.

What do you reckon?

File: docs/device_report.md (hunk 2 above)
Comment (JoelRedstone) [reply docs/device_report.md]: yeah my bad this time. Forgot to remove a carriage return

**Comment by JoelRedstone** [comment]
@InBrewJ PTAL

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+            String propertyName = map.getKey();
+            String propertyValue = map.getValue();
+            addObjectProperty(bacnetObjectType, propertyName, propertyValue);
+        }
+        addObjectType(localDevice, bacnetObjectType);
+    }
+
+    private void addObjectProperty(BACnetObject bacnetObjectType, String objectProperty, String propertyValue) {
+        Encodable encodable;
+        switch (objectProperty) {
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java:66]: @JoelRedstone The switch starts here

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+            case "Minimum_Off_Time":
+                minimumOffTime = Integer.parseInt(propertyValue);
+                encodable = new UnsignedInteger(minimumOffTime);
+                add(bacnetObjectType, PropertyIdentifier.minimumOffTime, encodable);
+                break;
+            case "Feeback_Value":
+                feedbackValue = Integer.parseInt(propertyValue);
+                encodable = new BinaryPV(feedbackValue);
+                add(bacnetObjectType, PropertyIdentifier.feedbackValue, encodable);
+                break;
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java:223]: @JoelRedstone the switch ends here

File: subset/bacnet/bacnetTests/.idea/gradle.xml [hunk 11]
```diff
@@ -15,4 +15,4 @@
       </GradleProjectSettings>
     </option>
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/.idea/gradle.xml:16]: @JoelRedstone I think these files are fairly redundant, see Trevor's cloud test:

https://github.com/grafnu/daq/tree/master/subset/cloud

And Ed's switch test:

https://github.com/arupiot/daq/tree/master/subset/switches

For what should be in the repo

Comment (InBrewJ) [reply subset/bacnet/bacnetTests/.idea/gradle.xml:16]: @JoelRedstone This is Trevor's pubber tool:

https://github.com/arupiot/daq/tree/master/pubber

...actually I could be off the mark about not including the `.idea` folder - maybe it includes important build stuff I don't know about?

File: subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java (hunk 7 above)
Comment (InBrewJ) [reply subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java:21]: Yep, an else would do the same thing and would require less changes to the existing code. Either or, just make sure everything is caught. 

Also:

> this is code _in the wild_, anything can happen

File: testing/test_aux.out [hunk 12]
```diff
@@ -6,9 +6,12 @@ Validator exit code 0
 Lint checks
 cmd/inbuild exit code 0
 Extended tests
-RESULT skip protocol.bacnet.version
-RESULT skip protocol.bacnet.version
+RESULT skip protocol.bacnet.version Bacnet device not found.
+RESULT skip protocol.bacnet.pic
 RESULT info protocol.bacnet.version Protocol version: 1
+RESULT fail protocol.bacnet.pic
+RESULT info protocol.bacnet.version Protocol version: 1
```
Comment (InBrewJ) [inline testing/test_aux.out:13]: @JoelRedstone Are you intentionally running the `info` result of `protocol.bacnet.version` twice?

**Review by InBrewJ — CHANGES_REQUESTED** [review]

File: docs/device_report.md [hunk 13]
```diff
@@ -127,7 +128,9 @@ RESULT fail connection.mac_oui
 ## Module bacext
 
 ```
-RESULT skip protocol.bacnet.version
+RESULT skip protocol.bacnet.version Bacnet device not found.
+RESULT skip protocol.bacnet.pic
+Bacnet device not found... Pics check cannot be performed.
```
Comment (InBrewJ) [inline docs/device_report.md]: @JoelRedstone is there another rogue carriage return here?

**Review by InBrewJ — CHANGES_REQUESTED** [review]

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java (hunk 4 above)
Comment (ewRedstone) [reply subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/Binary.java:219]: The default clause should throw an exception

File: testing/test_aux.out (hunk 12 above)
Comment (JoelRedstone) [reply testing/test_aux.out:13]: Well yes and no... bacnet.version and bacnet.pics test is performed for each faux device stated in "testing/test_aux.sh".

Once you enable the bacext module in Daq, it will run all bacnet related test for each device. 

https://github.com/JoelRedstone/daq/blob/bacnet_pics/subset/bacnet/test_bacext

In the link above you can see here that all tests are executed one after the other. 
These idea came from Trevor that said he wanted all bacnet test in a single test_*** file rather than having multiple test_*** and Dockerfile.test_*** file for each bacnet test (it would have been the same code in multiple files otherwise).

File: subset/bacnet/bacnetTests/.idea/gradle.xml (hunk 11 above)
Comment (JoelRedstone) [reply subset/bacnet/bacnetTests/.idea/gradle.xml:16]: I'll follow the pubber folder as an example.

File: docs/device_report.md (hunk 13 above)
Comment (JoelRedstone) [reply docs/device_report.md]: yep

**Comment by InBrewJ** [comment]
@JoelRedstone This is where I start to get confused as to why there are a load of new mac addresses in the /misc/test_site folder:

https://github.com/JoelRedstone/daq/tree/bacnet_pics/misc/test_site/mac_addrs

Travis is currently failing because there should be a device_id ("SNS-4") here:

https://github.com/grafnu/daq/blob/master/misc/test_site/mac_addrs/3c5ab41e8f0a/module_config.json

In your repo, you have a folder for a device with the same mac address:

https://github.com/JoelRedstone/daq/blob/bacnet_pics/misc/test_site/mac_addrs/3c5ab41e8f0a/module_config.json

which _does_ have a device_id ("SNS-4") - yet Travis is still failing on the udmi test because it can't find a device id.

So, questions:

- Why do we need extra devices in the mac_addr folder
- Is this making Travis fail?

Also, make sure this branch is up to date with grafnu/daq/master, things are changing fast in the background!

https://github.com/JoelRedstone/daq/tree/bacnet_pics/misc/test_site/mac_addrs

File: docs/device_report.md (hunk 13 above)
Comment (ewRedstone) [reply docs/device_report.md]: this same line description situation is not applicable to existing tests. project management advice

**Comment by ewRedstone** [comment]
@InBrewJ The mac addresses are all valid.
two google mac addresses and three faux mac addresses

It looks as though arupiot/daq Travis CI if failing on aux.  This would be a bad idea to merge arupiot/daq into this branch at this time.

File: testing/test_aux.out (hunk 12 above)
Comment (InBrewJ) [reply testing/test_aux.out:13]: Understood 👌

File: docs/device_report.md (hunk 13 above)
Comment (InBrewJ) [reply docs/device_report.md]: It seems to already be done for this test, but agreed

**Comment by InBrewJ** [comment]
@ewRedstone I understand that the mac addresses are valid - but why do we need them in the test_site?

I know there was talk about using the extra mac addresses for other tests but I can't remember what they were which is my bad. I'm just trying to make sure we're not adding things to the test_site that aren't used by Travis.

Also, test_aux failed on Travis because the GCP account it was connected to threw a fit for whatever reason. I re-ran the test and now it's fine. All good!

```
So, questions:

- Why do we need extra devices in the mac_addr folder (**still uncertain**)
- Is this making Travis fail? (**no**)
```

Shortened version:
More mac addresses here:
https://github.com/arupiot/daq/tree/master/misc/test_site/mac_addrs
Than there are here:
https://github.com/grafnu/daq/tree/master/misc/test_site/mac_addrs

Extra mac addresses introduced here:
https://github.com/arupiot/daq/commit/bd8b070fee2e30d1dcf319cf6275df2d9f51c1b9

**I know this is a commit that's already been merged**, but I'm still curious

File: subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    private String csvName = "pics";
+    private String csvExtension = ".csv";
+    private boolean debug = false;
+
+    public boolean checkDevicePicCSV() {
+        String csvFolder = getCSVPath();
+        try{
+        File[] listFiles = new File(csvFolder).listFiles();
+        for (int i = 0; i < listFiles.length; i++) {
+//            System.out.println(listFiles[i].getName());
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java]: Do we need this comment?

File: subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      } else { testPassed = propertyValidated; }
+    }
+    count++;
+  }
+
+  public Multimap<String, String> getResultMap() {
+    return result;
+  }
+
+  public boolean getRestult() {
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java]: @JoelRedstone typo: `getRestult`

File: subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/EntryPoint.java [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        String localIpAddr = args[0];
+        String broadcastIpAddr = args[1];
+        fauxDeviceJSONFilename = args[2];
+
+        int port = IpNetwork.DEFAULT_PORT;
+        network = new IpNetwork(broadcastIpAddr, port,
+                IpNetwork.DEFAULT_BIND_IP, 0, localIpAddr);
+        Transport transport = new Transport(network);
+        transport.setTimeout(1000);
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/FauxDeviceEngine/EntryPoint.java]: We could have a named constant for setTimeout too?

File: subset/bacnet/bacnetTests/src/main/java/helper/Csv.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+
+  public Csv(String csvFile) {
+    this.csvFile = csvFile;
+  }
+
+  public void readAndValidate(Multimap<String, Map<String, String>> bacnetPointsMap) {
+    try (BufferedReader br = new BufferedReader(new FileReader(csvFile))) {
+      while ((line = br.readLine()) != null) {
+        String[] value = line.split(csvSplitBy);
+        if (!value[0].equals(csvColumnTitle[0])
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/Csv.java]: @JoelRedstone I'd wager heavily that you'll look at this test again in 2 weeks and ask yourself:

- what is at position 0 in the value array?
- what is at position 1 in the value array?

I'd either write a comment or make 0/1/2/3 a named constant if you can

File: subset/bacnet/bacnetTests/src/main/java/PicsTest.java [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+public class PicsTest {
+
+  private Connection connection;
+  private BacnetValidation validator;
+  private BacnetPoints bacnetPoints = new BacnetPoints();
+  private String testName = "protocol.bacnet.pic";
+  private String passedTestReport = String.format("RESULT pass %s\n", testName);
+  private String failedTestReport = String.format("RESULT fail %s\n", testName);
+  private String skippedTestReport = String.format("RESULT skip %s", testName);
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/PicsTest.java]: I'm 80% sure we need a `\n` at the end of the skip line too

**Review by InBrewJ — CHANGES_REQUESTED** [review]

**Comment by ewRedstone** [comment]
It would be more helpful if you are able to submit all comments for review in one go.

This is a massive impact on development time and this is the key place time is getting lost.

In future please could you thoroughly work through the code and make all your comments in one swoop.

Otherwise what happens is we fix/challenge your comments and then re-submit.

Then "new" "issues" are found putting the test back into development and testing cycle.

It would be valuable if you please tighten down your process to avoid unnecessary loss of time.

File: subset/bacnet/bacnetTests/src/main/java/PicsTest.java (hunk 18 above)
Comment (JoelRedstone) [reply subset/bacnet/bacnetTests/src/main/java/PicsTest.java]: I've removed the "\n" for the description to be on the same line.

**Comment by ewRedstone** [comment]
> It would be more helpful if you are able to submit all comments for review in one go.
> 
> This is a massive impact on development time and this is the key place time is getting lost.
> 
> In future please could you thoroughly work through the code and make all your comments in one swoop.
> 
> Otherwise what happens is we fix/challenge your comments and then re-submit.
> 
> Then "new" "issues" are found putting the test back into development and testing cycle.
> 
> It would be valuable if you please tighten down your process to avoid unnecessary loss of time.

After our discussion today this makes sense that you are making the correct approach to the test development cycle. We need to communicate this to project management so they are aligned with our process.

File: subset/bacnet/bacnetTests/src/main/java/PicsTest.java (hunk 18 above)
Comment (InBrewJ) [reply subset/bacnet/bacnetTests/src/main/java/PicsTest.java]: @JoelRedstone  I think the issue is, if there are _no_ appendices found in the `appendicesMap` then the newline never gets added and it breaks the output markdown:

https://github.com/JoelRedstone/daq/blob/bacnet_pics/subset/bacnet/bacnetTests/src/main/java/helper/Csv.java#L90

You could add the newline here too? :

https://github.com/JoelRedstone/daq/blob/bacnet_pics/subset/bacnet/bacnetTests/src/main/java/helper/Csv.java#L93

File: subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+public class FileManager {
+
+    private String filePath = "";
+    private String csvName = "pics";
+    private String csvExtension = ".csv";
+    private boolean debug = false;
+
+    public boolean checkDevicePicCSV() {
+        String csvFolder = getCSVPath();
+        try{
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java:14]: [[indentation needed]]

File: subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+        File[] listFiles = new File(csvFolder).listFiles();
+        for (int i = 0; i < listFiles.length; i++) {
+            if (listFiles[i].isFile()) {
+                String fileName = listFiles[i].getName();
+                if (fileName.contains(csvName)
+                        && fileName.endsWith(csvExtension)) {
+                    System.out.println("pics.csv file found.");
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java]: This is the last thing I can see - can we print the path where the pics.csv file wasn't found? It makes debugging a lot easier for someone who isn't familiar with the system (i.e. all current users)

**Review by InBrewJ — CHANGES_REQUESTED** [review]

**Comment by JoelRedstone** [comment]
Hey @InBrewJ,
Can I ask you to double check the code and see if you have any more comments please? I would prefer receiving all comments at once rather than small bits (2-3comments at a time). It takes 2-3 hours to go through your comment, adjust the code, build and run daq to check if everything work, run testing_aux.sh script (twice) and then wait for Travis.

File: subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+                    String fileName = listFiles[i].getName();
+                    if (fileName.contains(csvName)
+                            && fileName.endsWith(csvExtension)) {
+                        System.out.println("pics.csv file found in /config/type.");
+                        setFilePath(fileName);
+                        return true;
+                    }
+                }
+            }
+            String errorMessage = "Pics.csv not found.\n";
```
Comment (InBrewJ) [inline subset/bacnet/bacnetTests/src/main/java/helper/FileManager.java]: @JoelRedstone The file path is also useful if it's printed when it's _not_ found

File: subset/connection/mac_oui/.classpath [hunk 22]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
 	<classpathentry kind="src" path="src/main/java"/>
-	<classpathentry kind="src" path="src/test/java"/>
+        <classpathentry kind="src" path="src/test/java"/>
```
Comment (InBrewJ) [inline subset/connection/mac_oui/.classpath]: @JoelRedstone Is this a garbage file?

File: subset/security/tlstest/.classpath [hunk 23]
```diff
@@ -6,11 +6,11 @@
 			<attribute name="gradle_used_by_scope" value="main,test"/>
 		</attributes>
 	</classpathentry>
-	<classpathentry kind="src" output="bin/test" path="src/test/java">
-		<attributes>
-			<attribute name="gradle_scope" value="test"/>
-			<attribute name="gradle_used_by_scope" value="test"/>
-		</attributes>
+        <classpathentry kind="src" output="bin/test" path="src/test/java">	
+		<attributes>
```
Comment (InBrewJ) [inline subset/security/tlstest/.classpath]: @JoelRedstone Is this a garbage file?

File: validator/.project [hunk 24]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <projectDescription>
-	<name>mac_oui</name>
-	<comment>Project mac_oui created by Buildship.</comment>
+	<name>validator</name>
```
Comment (InBrewJ) [inline validator/.project]: @JoelRedstone Is this a garbage file?

**Review by InBrewJ — CHANGES_REQUESTED** [review]

**Comment by ewRedstone** [comment]
> Hey @InBrewJ,
> Can I ask you to double check the code and see if you have any more comments please? I would prefer receiving all comments at once rather than small bits (2-3comments at a time). It takes 2-3 hours to go through your comment, adjust the code, build and run daq to check if everything work, run testing_aux.sh script (twice) and then wait for Travis.

Joel, I will highlight this at the project management meeting. I will let them decided how to tighten process.

File: subset/bacnet/bacnetTests/src/main/java/Main.java [hunk 25]
```diff
@@ -1,19 +1,22 @@
...
-    if (args.length < 3) {
-      throw new IllegalArgumentException("Usage: bacnetTestId broadcastIp localIp");
+    if (args.length < 4) {
+      throw new IllegalArgumentException("Usage: bacnetTestId broadcastIp localIp verbose-output");
     }
 
     String bacnetTestId = args[0];
     String broadcastIp = args[1];
     String localIp = args[2];
+    String verboseOutput = args[3];
```
Comment (pisuke) [inline subset/bacnet/bacnetTests/src/main/java/Main.java]: verboseOutput must be a boolean and changed in all relevant files

**Review by pisuke — COMMENTED** [review]
I'm doing the change to boolean so we can have this test approved and merged today

File: subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+        writeToAppendix(formatProperty, bacnetObjectType, bacnetObjectProperty, conformanceCode,
+                "PASSED", verboseOutput);
+        setResult(true);
+      }
+    }
+  }
+
+  private void writeToAppendix(String formatProperty, String key, String bacnetObjectProperty, String conformanceCode,
+                               String lineresult, boolean verboseOutput) {
+    if (!verboseOutput & (lineresult.contains("FAILED") | lineresult.contains("PASSED/WARNING"))) {
```
Comment (pisuke) [inline subset/bacnet/bacnetTests/src/main/java/helper/PicsValidator.java:70]: @JoelRedstone @ewRedstone @InBrewJ the code now makes use of a boolean verboseOutput switch which is false by default and only outputs FAILED and PASSED/WARNING fields/properties.
Let's fix Travis tomorrow and then merge.

**Review by pisuke — APPROVED** [review]
@JoelRedstone @ewRedstone @InBrewJ it's OK now from a minimal functional perspective, so we'll update the Travis test tomorrow and then merge.

**Comment by pisuke** [comment]
@InBrewJ the code and functionalities look good to me now. Could you please check that Travis doesn't fail and merge to arupiot/daq? Thanks!

**Comment by InBrewJ** [comment]
@pisuke A lot of the Travis builds are failing because various Maven packages can't be resolved, very odd but could be something to do with:

https://www.traviscistatus.com/

(at time of posting there was a partial outage on 'Linux and Windows builds')

Travis seems to be failing for bacnet_pics because of something to do with gRPC hanging though. I'm looking into it

**Comment by InBrewJ** [comment]
@pisuke Just done a few tests locally and daq runs everything fine, definitely connectivity from Travis' end

I realised I had to create a _new_ Travis build - hence why the commit below happened... ⬇️

**Comment by InBrewJ** [comment]
How Travis is failing for reference:

image

- port 1 is term (it was hanging on the ping test before disappearing)

Last lines of log before Travis gives up:

```
INFO:runner:Target port 3 finalize: [] (target set not active)
INFO:runner:Target port 3 cancel 3c:5a:b4:1e:8f:0a (#3/0).
INFO:network:Directing traffic for 3c:5a:b4:1e:8f:0a on port 3: False
INFO:topology:Waiting 5s for network to settle
INFO:host:Target port 3 terminate, trigger False
INFO:gcp:Unsubscribe callback device/3c:5a:b4:1e:8f:0a
INFO:google.cloud.firestore_v1.watch:RPC termination has signaled manager shutdown.
```

History of port-1 on the Travis dashboard:

image

Something is happening with ping...

**Comment by InBrewJ** [comment]
I've just noticed the following error in the Travis sandbox firebase account. It's happening very, very frequently:

```
TypeError: pubsub.topic(...).publisher is not a function
    at publishPubsubMessage (/srv/index.js:166:6)
    at exports.device_state.functions.pubsub.topic.onPublish (/srv/index.js:153:3)
    at cloudFunctionNewSignature (/srv/node_modules/firebase-functions/lib/cloud-functions.js:120:23)
    at /worker/worker.js:825:24
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)
```

which comes from here:

https://github.com/arupiot/daq/blob/master/firebase/functions/index.js#L166

the same line in grafnu:

https://github.com/grafnu/daq/blob/master/firebase/functions/index.js#L166

The same line in faucet:

https://github.com/faucetsdn/daq/blob/master/firebase/functions/index.js#L166

And Travis isn't failing in the same way on either grafnu _or_ faucet. This seems to be an entirely **weird** problem.

Running things locally isn't a problem, so publisher is clearly sometimes a function and sometimes isn't in a non-obvious way. Ah, Javascript.

While I figure this out I've taken the GCP_SERVICE_ACCOUNT credentials out of arupiot/daq. We're not testing anything to do with the cloud functionality - this should help Travis to actually do the job of testing bacnet_pics until this is sorted

**Comment by InBrewJ** [comment]
@pisuke @JoelRedstone , with Travis now _actually running_ the pics test:

image

The limit is allegedly 4MB: https://docs.travis-ci.com/user/common-build-problems/#log-length-exceeded

I suppose we can, of course, just increase the log size limit (if that's even possible). More realistically we might consider not outputting the entirety of the pics test into the log?

This would be a change in here to stop printing everything to `activate.log` I believe? :

https://github.com/JoelRedstone/daq/blob/bacnet_pics/subset/bacnet/bacnetTests/src/main/java/helper/Csv.java#L92

(it could be somewhere else, that was just after a first glance). And if that's a debug message, it probably shouldn't be there anyway. If we're printing out **thousands of lines** to anywhere (especially to a file or a virtual terminal on a Travis webpage) that's going to make running the test in Travis or on any machine running DAQ slower 🐌

**Comment by pisuke** [comment]
Thanks @InBrewJ yup, let's remove from the log. I removed outputting the PASSED lines, but having WARNING and FAILED should be useful. The WARNING should have also some information about why, but for the next time. I think we should act on the faux device to remove so many datapoints too.

**Review by InBrewJ — APPROVED** [review]
Travis is still being odd:
- failing to download Maven packages: https://travis-ci.org/arupiot/daq/jobs/569798705#L5664
- Erroring out on test_many sometimes (this is known - referenced in https://github.com/faucetsdn/daq/issues/74)
- etc

But the test works and the code is fine functionally and with faux containers, so approved!
